### ISSUE TITLE
Since v1.5.1 I'm getting "This test did not perform any assertions"

### ISSUE THREAD

**Issue by simPod** [mockery/mockery#1190]
I compared 1.5.1 to 1.5.0 and it seems the only function change is this https://github.com/mockery/mockery/pull/1180

I'm getting "This test did not perform any assertions" for tests that contain `->shouldReceive` calls.

It was counted towards assertions in all previouis Mockery versions.

**Comment by simPod** [comment]
And my main issue is that from the changelog I'm unable to find out what should I do.

**Comment by andrewbroberg** [comment]
Noticing the same issue here, could it be this change that has caused it potentially?

https://github.com/mockery/mockery/commit/72cf17bb9ac96fe52015c8fbc41b2a6b88e3177a

**Comment by andrewbroberg** [comment]
Found the reason why these are failing, it looks like it may be by design.

```php
Mockery::mock(Foo::class)->shouldReceive('bar');
```

The above will fail as we have not provided the number of calls we expect to it, we will need to update it to something like below

```php
Mockery::mock(Foo::class)->shouldReceive('bar')->once();
```

The added test in the commit show this expected behaviour https://github.com/mockery/mockery/commit/72cf17bb9ac96fe52015c8fbc41b2a6b88e3177a#diff-e290c245179d001a519df9ee5cecf659df2e2ff8506829ef10dcc6900559fe56

**Comment by simPod** [comment]
@andrewbroberg that's nice actually. I guess it's a fix then?

**Comment by ghostwriter** [comment]
correct, `->shouldReceive(…)` would result in zero assertions because it's only stubbing the method call.

**Comment by GrahamCampbell** [comment]
I think that is debatable. That should probably be regarded as a breaking change, and delayed till 2.x, and reverted in 1.x.

**Comment by davedevelopment** [comment]
Definitely debatable, but nobody turned up to the [debate](https://github.com/mockery/mockery/pull/1180#issuecomment-1137008454) and then I forgot...

**Comment by ghostwriter** [comment]
Regarding BC Break,

could/should we internally set the default expectation for method stubs `->shouldReceive(…)`  to `->once()` and allow it to be overwritten?

- `->shouldReceive(…)` would result in one assertion, addressing the issue that #1180 fixed.

``` php
$mock = Mockery::mock(); 
$mock->shouldReceive('foo'); // Same as: $mock->shouldReceive('foo')->once();
// can be overwritten
// $mock->shouldReceive('foo')->atLeast()->once();
// $mock->shouldReceive('foo')->never();

$this->assertSame(1, Mockery::getContainer()->mockery_getExpectationCount());
```

Thoughts?

**Comment by andrewbroberg** [comment]
IMO this is a good change, we should be calling `once` or similar otherwise we're not actually asserting on anything

**Comment by davedevelopment** [comment]
> Regarding BC Break,
> 
> could/should we internally set the default expectation for method stubs `->shouldReceive(…)` to `->once()` and allow it to be overwritten?
> ...
> Thoughts?

If what we've already done is considered a BC break, I would also consider that a BC break.

Regardless, this is essentially one of the driving forces for what I did for the (not so) new `allows` and `expects` syntax. `allows` for stubs, which is actually just a proxy to the old `shouldReceive` and `expects` for expectations, which is a proxy to `shouldReceive()->once()`.

```
    public function expects($something = null)              
    {                                                       
        if (is_string($something)) {                        
            return $this->shouldReceive($something)->once();
        }                                                   
                                                            
        return new ExpectsHigherOrderMessage($this);        
    }                                                       
```

**Comment by zlodes** [comment]
I thought that `shouldReceive` is deprecated 🤔 
It should be replaced with `expects` (means `shouldReceive` + `once`) and `allows` (just `shouldReceive`).

Because `should` word isn't fine there.

**Comment by mfn** [comment]
I've previously commented at https://github.com/mockery/mockery/pull/1180#issuecomment-1243280548 and only afterwards found this issue now.

Turns out the change is affecting us in a bigger way than I initially was able to see.

From a testsuite of 20k+ tests, we have ~500+ being reported as "Risky".

What added more to the confusion in our case is how different environments treated this.

- phpunit (9.5.25), with default `beStrictAboutTestsThatDoNotTestAnything=true`, will report them in test runs but will still return exit code 0 => that's why I initially didn't "see" the 500+ because the GHA workflow runs were green
- however we've split teststuites in phpunit and run some through paratest. paratest however will return a non-0 exit code when encountering risky tests with above settings => this is how I initially only found 1 test with a dataprovider having "an issue"
  Apparently the [paratest behaviour is due to the junit log format](https://github.com/paratestphp/paratest/issues/282#issuecomment-715254329), if that is still the most accurate information on that.
- also in phpstorm, which uses phpunits `--teamcity` flag, those risky tests are marked as `testFailed` and thus they appear as failed, when in phpunit by default they wouldn't.
  Interestingly, phpstorm could check phpunits exit code because even with --teamcity it's `0` but it honors only the `testFailed` status there.
  But we didn't find this out until later, because the 20k+ tests are mostly integration/feature tests and it's usually not feasible to run them locally.

---
However we also agree those tests are broken and need to be fixed, so we're not advocating for any change here (1.5.1 is released, aka "damage is done" anyway) and are simply going ahead and add a lot of `expectNotToPerformAssertions()` calls where previously none were 🤞🏼

**Comment by jrfnl** [comment]
Just wanted to point out that this doesn't only affect userland tests, but also the tests of Mockery itself.

I've [re-run the tests for Mockery itself on the commit before PR #1180 was merged](https://github.com/jrfnl/mockery/actions/runs/3091583499/jobs/5001831593) (as GH Actions had already thrown away the logs): this shows 5 tests which are marked as "risky" by PHPUnit (prior to the change).

When we compare this to a [(re-run) build of the commit merging PR #1180](https://github.com/jrfnl/mockery/actions/runs/3091652050), it shows between 39 and 44 tested marked as "risky" by PHPUnit (exact number depends on the PHP version used).

**Comment by mfn** [comment]
> phpunit (9.5.25), with default `beStrictAboutTestsThatDoNotTestAnything=true`, will report them in test runs but will still return exit code 0 => that's why I initially didn't "see" the 500+ because the GHA workflow runs were green

FTR, no idea how I came up with this strange setting but simply adding `failOnRisky="true"` made them a hard stop.

We also found a couple of bugs in our test suite thanks to this change 🙏🏼

**Comment by downsider** [comment]
While not debating that this was a bug that needed fixing, the 1.5.1 patch change has caused our test suite to throw several new failure cases. That's a breaking change in no uncertain terms.

Semver would suggest that this change be delayed until the next major release. To leave it as-is has a negative impact on downstream users; this could be preventing someone from deploying an urgent bugfix to their application. Regardless of whether a test suite is technically flawed or not, we shouldn't be forcing people to update their tests just because they installed a new patch version of mockery.

For the record, this kind of change is why I prefer using [Ferver](https://github.com/jaredly/ferver) over Semver. Breaking changes should not be something to be afraid of, you just have to ensure that people don't get caught out by them. In Ferver, this would be a new minor version; if people wanted to update to the latest version, they would need to consciously change composer.json and can handle the consequences of that appropriately.

> we'll update the changelog to reflect the changes.

Seeing as this is outstanding, please see #1192

**Comment by BenMorel** [comment]
> IMO this is a good change, we should be calling once or similar otherwise we're not actually asserting on anything

The issue is, if you have other assertions in your test, then PHPUnit will happily mark the test as successful, even though the expectation failed:

```php
interface I {
    public function test(): void;
}

class MyTest extends TestCase
{
    public function test(): void
    {
        $i = Mockery::mock(I::class);
        $i->shouldReceive('test');

        self::assertTrue(true);

        Mockery::close(); // will not complain, and PHPUnit will mark the test as successful!
    }
}
```

Calling `shouldReceive()` alone is akin to doing nothing, which can hide potential bugs as you can see here.

I can see a couple ways to solve this:

- as suggested by @ghostwriter [above](https://github.com/mockery/mockery/issues/1190#issuecomment-1241136610):
    > set the default expectation for method stubs ->shouldReceive(…) to ->once() and allow it to be overwritten
- or, store a flag in `Container` to mark `shouldReceive()` as unconfigured until `once()`/`never()`/... has been called, and throw in `Mockery::close()` if this flag is set

Also, if `shouldReceive()` is deprecated in favour of `expects()`, it would be nice to document it as `@deprecated`, so that we can at least run static analysis on our tests to detect potential issues.

Thoughs?

**Comment by Danny-Smart** [comment]
> set the default expectation for method stubs ->shouldReceive(…) to ->once() and allow it to be overwritten

We often use a blank `shouldReceive()` to prevent unexpected invocation errors where it is not possible or desirable to use `shouldIgnoreMissing()`. This happens frequently where loggers are used, or for other cases where a `void` return method is called that has nothing to do with the behaviour under test.

If we did go down this route, I think we would also need to add an `->any()` method, to allow the previous behaviour to be preserved

```php
$mock = Mockery::mock(Example::class);
// Currently, the two following expectations would be identical
// If we assign `->once()` by default, the latter allows for method calls we don't need to worry about in this test
$mock->shouldReceive('test');
$mock->shouldReceive('test')->any();
```
